blocks_reexecutor/blocks_reexecutor.go

```diff
@@ -111,32 +123,47 @@ func New(c *Config, blockchain *core.BlockChain, fatalErrChan chan error) *Block
 	if start > 0 && start != chainStart {
 		start--
 	}
-	// Divide work equally among available threads when BlocksPerThread is zero
-	if c.BlocksPerThread == 0 {
+	// Divide work equally among available threads when MinBlocksPerThread is zero
+	if c.MinBlocksPerThread == 0 {
 		// #nosec G115
 		work := (end - start) / uint64(c.Room)
 		if work > 0 {
-			blocksPerThread = work
+			minBlocksPerThread = work
 		}
 	}
-	return &BlocksReExecutor{
-		config:          c,
-		blockchain:      blockchain,
-		currentBlock:    end,
-		startBlock:      start,
-		blocksPerThread: blocksPerThread,
-		done:            make(chan struct{}, c.Room),
-		fatalErrChan:    fatalErrChan,
-		stateFor: func(header *types.Header) (*state.StateDB, arbitrum.StateReleaseFunc, error) {
-			state, err := blockchain.StateAt(header.Root)
-			return state, arbitrum.NoopStateRelease, err
-		},
+	hashConfig := *hashdb.Defaults
+	hashConfig.CleanCacheSize = c.TrieCleanLimit * 1024 * 1024
+	trieConfig := triedb.Config{
+		Preimages: false,
+		HashDB:    &hashConfig,
+	}
+	blocksReExecutor := &BlocksReExecutor{
+		config:             c,
+		db:                 state.NewDatabaseWithConfig(ethDb, &trieConfig),
+		blockchain:         blockchain,
+		currentBlock:       end,
+		startBlock:         start,
+		minBlocksPerThread: minBlocksPerThread,
+		done:               make(chan struct{}, c.Room),
+		fatalErrChan:       fatalErrChan,
+	}
+	blocksReExecutor.stateFor = func(header *types.Header) (*state.StateDB, arbitrum.StateReleaseFunc, error) {
+		sdb, err := state.NewDeterministic(header.Root, blocksReExecutor.db)
+		if err == nil {
+			_ = blocksReExecutor.db.TrieDB().Reference(header.Root, common.Hash{}) // Will be dereferenced later in advanceStateUpToBlock
+			stateReleaseFunc := func() {
+				_ = blocksReExecutor.db.TrieDB().Dereference(header.Root)
+			}
+			return sdb, stateReleaseFunc, nil
+		}
+		return sdb, arbitrum.NoopStateRelease, err
 	}
+	return blocksReExecutor, nil
 }
 
-// LaunchBlocksReExecution launches the thread to apply blocks of range [currentBlock-s.config.BlocksPerThread, currentBlock] to the last available valid state
+// LaunchBlocksReExecution launches the thread to apply blocks of range [currentBlock-s.config.MinBlocksPerThread, currentBlock] to the last available valid state
 func (s *BlocksReExecutor) LaunchBlocksReExecution(ctx context.Context, currentBlock uint64) uint64 {
```

Comment:
this still has a problem with re-execution.
Say we want to reexecute blocks 1000-2000, with 4 thread, and we have in database only the state for blocks [1000, 1205, 1405, 1605]:
Current algorithm will launch 4 threads: A for blocks 1000-1200, B for blocks 1200-1400 and so forth.
But when thread B tries to reexecute block 1200 - there is no state, so it has to reexecute from block 1000 all the way to 1200 and then to 1400. Thread C and D will be similar.
End result: Most blocks were computed twice, wasting a lot of resources (and if 1405 isn't in DB - many will be computed 3 times)
I described in linear an algorithm where we'll only launch threads after finding the state in DB.
So if we have MinBlocksPerThread as 200 - we'll end up with 3 threads, one computing state 1000-1205 one 1205-1405 and one 1405-2000. It's less parallelization - but we only compute each block once. No work is wasted.
In this algorithm, it makes sense (for this reason) to set MinBlocksPerThread lower than #blocks/#threads, and if we do that we'll get 4 threads in the scenario I described.

Reply:
In the current implementation we would first start the re-execution with end block as 2000 and look for a state at `start = 2000-minBlocksPerThread` (As minBlocksPerThread is already calculated as (end-start)/room so it would be 1000/4=250) -> which will give us start at 1750 and we look for state at this block -> we find 1605 and set re-execution from 1605 to 2000, but note that we set s.currentBlock to this start i.e 1605 meaning, next thread will start with range's end block as 1605 implying there would be no block that gets re-executed twice.

So the re-execution ranges would be 
* 1605-2000
* 1205-1605
* 1000-1205